op/main.py

```diff
@@ -116,10 +116,25 @@ def _get_event_args(charm, bound_event):
         relation_name = os.environ['JUJU_RELATION']
         relation_id = int(os.environ['JUJU_RELATION_ID'].split(':')[-1])
         relation = model.get_relation(relation_name, relation_id)
-        if issubclass(event_type, op.charm.RelationUnitEvent):
-            remote_unit_name = os.environ['JUJU_REMOTE_UNIT']
-            remote_unit = model.get_unit(remote_unit_name)
-            return [relation, remote_unit], {}
+
+        remote_app_name = os.environ.get('JUJU_REMOTE_APP')
+        remote_unit_name = os.environ['JUJU_REMOTE_UNIT']
+
+        if remote_app_name is None:
+            if remote_unit_name:
+                remote_app_name = remote_unit_name.split('/')[0]
+
+        if remote_app_name:
+            remote_app = model.get_app(remote_app_name)
+        else:
+            remote_app = None
+
+        if remote_app:
+            if remote_unit_name:
+                remote_unit = model.get_unit(remote_unit_name)
+                return [relation, remote_app, remote_unit], {}
+            else:
+                return [relation, remote_app], {}
```

Comment:
The logic above seems quite complicated, going back and forth between the two variables several times, but it looks like the conditions are much more strict than these branches make it sound.

Either we have a unit or we don't. That's about it. So this ought to be much simpler:
```
remote_app_name = os.environ.get('JUJU_REMOTE_APP', '')
remote_unit_name = os.environ.get('JUJU_REMOTE_UNIT', '')
if remote_app_name or remote_unit_name:
        if not remote_app_name:
                if '/' not in remote_unit_name:
                        raise RuntimeError(f'invalid remote unit name: {remote_unit_name}')
                remote_app_name = remote_unit_name.split('/')[0]
        args = [relation, model.get_app(remote_app_name)]
        if remote_unit_name:
                args.append(model.get_unit(remote_unit_name))
        return args, {}
return [], {}
```
The exception type is incorrect, but we can tune that later.

Reply:
OK, there is one missing bit in the above which is handling of `broken` relations - a relation needs to be returned without any units or applications present in that case. I will add it to my commit.